primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -224,326 +172,213 @@ impl OverlayedChanges {
 	/// to the backend); Some(None) if the key has been deleted. Some(Some(...)) for a key whose
 	/// value has been set.
 	pub fn storage(&self, key: &[u8]) -> Option<Option<&[u8]>> {
-		self.prospective.top.get(key)
-			.or_else(|| self.committed.top.get(key))
-			.map(|x| {
-				let size_read = x.value.as_ref().map(|x| x.len() as u64).unwrap_or(0);
-				self.stats.tally_read_modified(size_read);
-				x.value.as_ref().map(AsRef::as_ref)
-			})
+		self.top.get(key).map(|x| {
+			let value = x.value();
+			let size_read = value.map(|x| x.len() as u64).unwrap_or(0);
+			self.stats.tally_read_modified(size_read);
+			value.map(AsRef::as_ref)
+		})
 	}
 
-	/// Returns mutable reference to current changed value (prospective).
-	/// If there is no value in the overlay, the default callback is used to initiate
-	/// the value.
-	/// Warning this function register a change, so the mutable reference MUST be modified.
+	/// Returns mutable reference to current value.
+	/// If there is no value in the overlay, the default callback is used to initiate the value.
+	/// Warning this function registers a change, so the mutable reference MUST be modified.
+	///
+	/// Can be rolled back or committed when called inside a transaction.
 	#[must_use = "A change was registered, so this value MUST be modified."]
 	pub fn value_mut_or_insert_with(
 		&mut self,
 		key: &[u8],
 		init: impl Fn() -> StorageValue,
 	) -> &mut StorageValue {
-		let extrinsic_index = self.extrinsic_index();
-		let committed = &self.committed.top;
-
-		let mut entry = self.prospective.top.entry(key.to_vec())
-			.or_insert_with(|| {
-				if let Some(overlay_state) = committed.get(key).cloned() {
-					overlay_state
-				} else {
-					OverlayedValue { value: Some(init()), ..Default::default() }
-				}
-			});
-
-		//if was deleted initialise back with empty vec
-		if entry.value.is_none() {
-			entry.value = Some(Default::default());
-		}
-		if let Some(extrinsic) = extrinsic_index {
-			entry.extrinsics.get_or_insert_with(Default::default)
-				.insert(extrinsic);
+		let value = self.top.modify(key.to_owned(), init, self.extrinsic_index());
+
+		if value.is_none() {
+			*value = Some(Default::default());
```

Comment:
With the same argument you could abolish all use of type inference. Your tooling will tell you the type.

Reply:
Unless your tooling won't show the type, as it is for those poor souls who don't use IDEs or who read this from diffs : )

Also I disagree with your statement: type inference usually saves a lot of characters where (while writing and most importantly reading), whereas here it doesn't really matter since the difference is minor while improving the readability.

Feel free to ignore me here though. I don't feel that it is very important one.